sys-utils/libcmt/src/rollup.c

```diff
@@ -0,0 +1,401 @@
+    ||  DBG(cmt_abi_get_uint(rd, sizeof(advance->index), &advance->index))
+    ||  DBG(cmt_abi_get_bytes_s(rd, of))
+    ||  DBG(cmt_abi_get_bytes_d(anchor, of, &payload_length, &advance->payload))) {
+        return -ENOBUFS;
+    }
+    // clang-format on
+
+    advance->payload_length = payload_length;
+    return 0;
+}
+
+int cmt_rollup_read_inspect_state(cmt_rollup_t *me, cmt_rollup_inspect_t *inspect) {
+    if (!me) {
+        return -EINVAL;
+    }
+    if (!inspect) {
+        return -EINVAL;
+    }
+
+    cmt_buf_t rd[1];
+    if (DBG(cmt_rollup_get_rx(me, rd))) {
+        return -ENOBUFS;
+    }
+
+    inspect->payload_length = cmt_buf_length(rd);
+    inspect->payload = rd->begin;
+    return 0;
+}
+
+static int accepted(union cmt_io_driver *io, uint32_t *n) {
+    struct cmt_io_yield req[1] = {{
+        .dev = HTIF_DEVICE_YIELD,
+        .cmd = HTIF_YIELD_CMD_MANUAL,
+        .reason = HTIF_YIELD_MANUAL_REASON_RX_ACCEPTED,
+        .data = *n,
+    }};
+    int rc = DBG(cmt_io_yield(io, req));
+    if (rc) {
+        return rc;
+    }
+
+    *n = req->data;
+    return req->reason;
+}
+
+static int revert(union cmt_io_driver *io) {
+    struct cmt_io_yield req[1] = {{
+        .dev = HTIF_DEVICE_YIELD,
+        .cmd = HTIF_YIELD_CMD_MANUAL,
+        .reason = HTIF_YIELD_MANUAL_REASON_RX_REJECTED,
+        .data = 0,
+    }};
+    return DBG(cmt_io_yield(io, req));
+}
+
+int cmt_rollup_finish(cmt_rollup_t *me, cmt_rollup_finish_t *finish) {
+    if (!me) {
+        return -EINVAL;
+    }
+    if (!finish) {
+        return -EINVAL;
+    }
+
+    if (!finish->accept_previous_request) {
+        return revert(me->io); /* revert should not return! */
+    }
+
+    cmt_merkle_get_root_hash(me->merkle, cmt_io_get_tx(me->io).begin);
+    me->fromhost_data = CMT_WORD_LENGTH;
+    int reason = accepted(me->io, &me->fromhost_data);
+    if (reason < 0) {
+        return reason;
+    }
+    finish->next_request_type = reason;
+    finish->next_request_payload_length = me->fromhost_data;
+    cmt_merkle_init(me->merkle);
+    return 0;
+}
+
+int cmt_gio_request(cmt_rollup_t *me, cmt_gio_t *req) {
+    if (!me) {
+        return -EINVAL;
+    }
+    if (!req) {
+        return -EINVAL;
+    }
+
+    /* Accept empty id */
+    if (!req->id && req->id_length) {
+        return -EINVAL;
+    }
+
+    cmt_buf_t wr[1] = {cmt_io_get_tx(me->io)};
+    cmt_buf_t _[1];
+    if (cmt_buf_split(wr, req->id_length, wr, _)) {
+        return -ENOBUFS;
+    }
+    if (req->id) {
+        memcpy(wr->begin, req->id, req->id_length);
+    }
+
+    struct cmt_io_yield rr[1] = {{
+        .dev = HTIF_DEVICE_YIELD,
+        .cmd = HTIF_YIELD_CMD_MANUAL,
+        .reason = req->domain,
+        .data = req->id_length,
+    }};
+
+    int rc = DBG(cmt_io_yield(me->io, rr));
+    if (rc != 0) {
+        return rc;
+    }
+
+    cmt_buf_t rd[1];
+    if (cmt_rollup_get_rx(me, rd)) {
+        return -ENOBUFS;
+    }
+
+    size_t rd_length = cmt_buf_length(rd);
```

Comment:
This check is now redundant since `cmt_rollup_get_rx` will fail if data is larger than buffer.